README.md
Rack::Defense
=============

A Rack middleware for throttling and filtering requests.

[![Build Status](https://travis-ci.org/nakhli/rack-defense.svg)](https://travis-ci.org/nakhli/rack-defense)
[![Security](https://hakiri.io/github/nakhli/rack-defense/master.svg)](https://hakiri.io/github/nakhli/rack-defense/master)
[![Code Climate](https://codeclimate.com/github/nakhli/rack-defense/badges/gpa.svg)](https://codeclimate.com/github/nakhli/rack-defense)
[![Gem Version](https://badge.fury.io/rb/rack-defense.svg)](http://badge.fury.io/rb/rack-defense)

Rack::Defense is a Rack middleware that allows to easily add request rate limiting and request filtering to your Rack based application (Ruby On Rails, Sinatra etc.).

* [Request throttling](#throttling) (aka rate limiting) happens on __sliding window__ using the provided period, request criteria and maximum request number. It uses Redis to track the request rate.

* [Request filtering](#filtering) bans (rejects) requests based on provided criteria.

Rack::Defense has a small footprint and only two dependencies: [rack](https://github.com/rack/rack) and [redis](https://github.com/redis/redis-rb).

Rack::Defense is inspired from the [Rack::Attack](https://github.com/kickstarter/rack-attack) project. The main difference is the throttling algorithm: Rack::Attack uses a counter reset at the end of each period, therefore allowing up to 2 times more requests than the maximum rate specified. We use a sliding window algorithm allowing a precise request rate limiting.

## Getting started

Install the rack-defense gem; or add it to you Gemfile with bundler:

```ruby
# In your Gemfile
gem 'rack-defense'
```

Tell your app to use the Rack::Defense middleware. For Rails 3+ apps:

```ruby
# In config/application.rb
config.middleware.use Rack::Defense
```

Or for Rackup files:

```ruby
# In config.ru
use Rack::Defense
```

Add a `rack-defense.rb` file to `config/initializers/`:

```ruby
# In config/initializers/rack-defense.rb
Rack::Defense.setup do |config|
  # your configuration here
end
```

## Throttling

The Rack::Defense middleware evaluates the throttling criteria (lambdas) against the incoming request.
If the return value is falsy, the request is not throttled. Otherwise, the returned value is used as a key to
throttle the request. The returned key could be the request IP, user name, API token or any discriminator to throttle
the requests against.

### Examples

Throttle POST requests for path `/login` with a maximum rate of 3 request per minute per IP:

```ruby
Rack::Defense.setup do |config|
  config.throttle('login', 3, 60 * 1000) do |req|
    req.ip if req.path == '/login' && req.post?
  end
end
```

Throttle GET requests for path `/api/*` with a maximum rate of 50 request per second per API token:

```ruby
Rack::Defense.setup do |config|
  config.throttle('api', 50, 1000) do |req|
    req.env['HTTP_AUTHORIZATION'] if %r{^/api/} =~ req.path
  end 
end
```

Throttle POST requests for path `/aggregate/report` with a maximum rate of 10 requests per hour for a given logged in user. We assume here that we are using the [Warden](https://github.com/hassox/warden) middleware for authentication or any Warden based authentication wrapper, like [Devise](https://github.com/plataformatec/devise) in Rails.

```ruby
Rack::Defense.setup do |config|
  config.throttle('aggregate_report', 10, 1.hour.in_milliseconds) do |req|
    req.env['warden'].user.id if req.path == '/aggregate/report' && req.env['warden'].user
  end 
end
```

### Redis Configuration

Rack::Defense uses Redis to track request rates. By default, the `REDIS_URL` environment variable is used to setup
the store. If not set, it falls back to host `127.0.0.1` port `6379`.
The redis store can be setup with either a connection url: 

```ruby
Rack::Defense.setup do |config|
  config.store = "redis://:p4ssw0rd@10.0.1.1:6380/15"
end
```

or directly with a connection object:

```ruby
Rack::Defense.setup do |config|
  config.store = Redis.new(host: "10.0.1.1", port: 6380, db: 15)
end
```

## Filtering

Rack::Defense can reject requests based on arbitrary properties of the request. Matching requests are filtered out.

### Examples

Allow only a whitelist of IPs for a given path:

```ruby
Rack::Defense.setup do |config|
  config.ban('ip_whitelist') do |req|
    req.path == '/protected' && !['192.168.0.1', '127.0.0.1'].include?(req.ip)
  end
end
```

Deny access to a blacklist of application users. Again, we assume here that 
[Warden](https://github.com/hassox/warden) or any Warden based authentication wrapper, like [Devise](https://github.com/plataformatec/devise), is used:

```ruby
Rack::Defense.setup do |config|
  config.ban('user_blacklist') do |req|
    ['hacker@example.com', 'badguy@example.com'].include? req.env['warden'].user.email
  end 
end
```

Allow only requests with a known API authorization token:

```ruby
Rack::Defense.setup do |config|
  config.ban('validate_api_token') do |req|
    %r{^/api/} =~ req.path && Redis.current.sismember('apitokens', req.env['HTTP_AUTHORIZATION'])
  end
end
```

The previous example uses redis to keep track of valid api tokens, but any store (database, key-value store etc.) would do here.

## Response configuration

By default, Rack::Defense returns `429 Too Many Requests` and `403 Forbidden` respectively for throttled and banned requests.
These responses can be fully configured in the setup:

```ruby
Rack::Defense.setup do |config|
  config.banned_response =
    ->(env) { [404, {'Content-Type' => 'text/plain'}, ["Not Found\n"]] }
  config.throttled_response =
    ->(env) { [503, {'Content-Type' => 'text/plain'}, ["Service Unavailable\n"]] }
end
```

## Notifications

You can be notified when requests are throttled or banned. The callback receives the throttled request object and data
about the event context.

For banned request callbacks, the triggered rule name is passed: 

```ruby
Rack::Defense.setup do |config|
  config.after_ban do |req, rule|
    logger.info "[Banned] #{rule} #{req.path} #{req.ip}"
  end
end
```

For throttled request callbacks, a hash having triggered rule names as keys and the corresponding throttle keys
as values is passed. 

```ruby
Rack::Defense.setup do |config|
  config.after_throttle do |req, rules|
    logger.info rules.map { |e| "[Throttled] rule name: #{e[0]} - rule throttle key: #{e[1]}" }.join ', '
  end
end
```

## Advanced Examples

### Temporarily suspend access to suspicious IPs

In this example, when an IP is exceeding the permitted request rate, we would like to ban this IP for a given period of time:

```ruby
Rack::Defense.setup do |config|
  config.throttle('reset_password', 10, 10.minutes.in_milliseconds) do |req|
    req.ip if req.path == '/api/users/password' && req.post?
  end

  config.after_throttle do |req, rules|
    config.store.setex("ban:ip:#{req.ip}", 1.hour, 1) if rules.key? 'reset_password'
  end

  config.ban('blacklist') do |req|
    config.store.exists("ban:ip:#{req.ip}")
  end
end
```

The first rule named `reset_password` defines the maximum permitted rate per IP for post requests on path
`/api/users/password`. Once a user exceeds this limit, it gets throttled and denied access to the resource.
This raises a throttle event and triggers the `after_throttle` callback defined above. The callback sets a key in the redis store post-fixed with the user IP and having 1 hour an expiration time.

The last rule named `blacklist` looks up each incoming request IP and checks if it has a corresponding ban key
in redis. If the request IP matches a ban key it gets denied.

## License

Licensed under the [MIT License](http://opensource.org/licenses/MIT).

Copyright Chaker Nakhli.



CHANGELOG.md
# 0.2.6
tag 0.2.6

Update rack version due to vulnerability
Update redis version

# 0.2.5
tag 0.2.5

Update dependencies.

# 0.2.4
tag 0.2.4

Update dependencies and author info in gemspec.

# 0.2.3
tag 0.2.3

Update dependencies.

# 0.2.2
tag 0.2.2

Update dependencies.

# 0.2.1
tag 0.2.1

* Added auto-expiration of redis throttle keys using the redis command [PEXPIRE](http://redis.io/commands/pexpire).
A throttle key is automatically cleaned up by redis if no activity is recorded against this key for more
than the specified throttle period.

# 0.2.0
tag 0.2.0

* Added notifications for throttle and ban events. Callbacks are registered with `Config#after_ban` and
`Config#after_throttle` methods.

# 0.1.1
tag 0.1.1

* Relax rack and redis gem required versions
* Wrap redis connection with a proxy before initializing `ThrottleCounter` instances.
This avoids having to do the store initialization (`Config#sotre=`) -if any- before declaring
throttle rules (`Config#throttle`). For instance, the following configuration is now correct:

```ruby
Rack::Defense.setup do |config|
  config.throttle('name', 100, 1000) { |req| req.ip if req.path='/path' }

  # no need to set the store before the throttle rule. it can be done at any moment in config section
  config.store = 'redis://server:3333/0'
end
```

# 0.1.0
tag 0.1.0

* Throttle requests using a sliding window with period/max_request and request criteria.
* Ban (block) requests matching criteria.
* `Rack::Defense#setup` to configure redis store and throttled and banned responses


spec/spec_helper.rb
require 'minitest/autorun'
require 'rack/test'
require 'redis'
require 'timecop'
require 'rack/defense'

class MiniTest::Spec
  include Rack::Test::Methods

  def status_ok; 200 end
  def status_throttled; 429 end
  def status_banned; 403 end

  def app
    Rack::Builder.new {
      use Rack::Defense
      run ->(_) { [200, {}, ['Hello World']] }
    }.to_app
  end

  before do
    Timecop.safe_mode = true
    keys = Redis.current.keys("#{Rack::Defense::ThrottleCounter::KEY_PREFIX}:*")
    Redis.current.del *keys if keys.any?
  end
end


spec/defense_spec.rb
require_relative 'spec_helper'

describe 'Rack::Defense::throttle' do
  STATUS_OK = 200
  STATUS_THROTTLED = 429
  PERIOD = 60 * 1000 # in milliseconds

  before do
    @start_time = Time.utc(2015, 10, 30, 21, 0, 0)

    #
    # configure the Rack::Defense middleware with two throttling
    # startegies.
    #
    Rack::Defense.setup do |config|
      # allow only 3 post requests on path '/login' per PERIOD per ip
      config.throttle('login', 3, PERIOD) do |req|
        req.ip if req.path == '/login' && req.post?
      end

      # allow only 50 get requests on path '/search' per PERIOD per ip
      config.throttle('res', 30, PERIOD) do |req|
        req.ip if req.path == '/search' && req.get?
      end
    end
  end
  it 'allow ok post' do
    check_post_request
  end
  it 'allow ok get' do
    check_get_request
  end
  it 'ban get requests higher than acceptable rate' do
    10.times do |period|
      50.times { |offset| check_get_request(offset + period*PERIOD) }
    end
  end
  it 'ban post requests higher than acceptable rate' do
    10.times do |period|
      7.times { |offset| check_post_request(offset + period*PERIOD) }
    end
  end
  it 'not have side effects between differrent throttle rules with mixed requests' do
    10.times do |period|
      50.times do |offset|
        check_get_request(offset + period*PERIOD)
        check_post_request(offset + period*PERIOD)
      end
    end
  end
  it 'not have side effects between request filtered by the same rule but with different keys' do
    10.times do |period|
      50.times do |offset|
        check_get_request(offset + period*PERIOD, ip='192.168.0.1')
        check_get_request(offset + period*PERIOD, ip='192.168.0.2')
      end
    end
  end
  it 'allow unfiltered requests' do
    50.times do |offset|
      time = @start_time + offset
      Timecop.freeze(time) do
        # the rule matches the '/search' path and not '/searchx'
        get '/searchx', {}, 'REMOTE_ADDR' => '192.168.0.1'
        assert_equal STATUS_OK, last_response.status

        # the rule matches only get requests and not post
        post '/search', {}, 'REMOTE_ADDR' => '192.168.0.1'
        assert_equal STATUS_OK, last_response.status
      end
    end
    10.times do |offset|
      time = @start_time + offset
      Timecop.freeze(time) do
        # the rule matches only post and not get
        get '/login', {}, 'REMOTE_ADDR' => '192.168.0.1'
        assert_equal STATUS_OK, last_response.status
      end
    end
  end
  it 'not have side effects between unfiltered and filtered requests' do
    50.times do |offset|
      time = @start_time + offset
      Timecop.freeze(time) do
        get '/searchx', {}, 'REMOTE_ADDR' => '192.168.0.1'
        assert_equal STATUS_OK, last_response.status
        get '/search', {}, 'REMOTE_ADDR' => '192.168.0.1'
        assert_equal offset < 30 ? STATUS_OK : STATUS_THROTTLED, last_response.status
      end
    end
  end

  def check_get_request(time_offset=0, ip='192.168.0.1', path='/search')
    check_request(:get, path, time_offset, 30, ip)
  end

  def check_post_request(time_offset=0, ip='192.168.0.1', path='/login')
    check_request(:post, path, time_offset, 3, ip)
  end

  def check_request(verb, path, time_offset, max_requests, ip)
    Timecop.freeze(@start_time + time_offset) do
      send verb, path, {}, 'REMOTE_ADDR' => ip
      expected_status = (time_offset % PERIOD) >= max_requests ? STATUS_THROTTLED : STATUS_OK
      assert_equal expected_status, last_response.status, "offset #{time_offset}"
    end
  end
end



spec/throttle_spec.rb
require_relative 'spec_helper'

describe 'Rack::Defense::Throttle' do
  before do
    Redis.current.flushdb
    @counter = Rack::Defense::Throttle.new('upload_photo', 5, 10, Redis.current)
    @key = '127.0.0.1'
  end

  describe '.throttle?' do
    it 'allow request number max_requests if after period' do
      do_max_requests_minus_one
      refute @counter.throttle? @key, 11
    end
    it 'block request number max_requests if in period' do
      do_max_requests_minus_one
      assert @counter.throttle? @key, 10
    end
    it 'allow consecutive valid periods' do
      (0..20).each { |i| do_max_requests_minus_one(11 * i) }
    end
    it 'block consecutive invalid requests' do
      do_max_requests_minus_one
      (0..20).each { |i| assert @counter.throttle?(@key, 10 + i) }
    end
    it 'use a sliding window and not reset count after each full period' do
      [6, 7, 8, 9].each { |t| refute @counter.throttle?(@key, t), "timestamp #{t}" }
      [12, 13, 14].each { |t| assert @counter.throttle?(@key, t), "timestamp #{t}"}
    end
    it 'should unblock after blocking requests' do
      do_max_requests_minus_one
      assert @counter.throttle? @key, 10
      assert @counter.throttle? @key, 11
      refute @counter.throttle? @key, 16
    end
    it 'should include throttled(blocked) request into the request count' do
      [0, 1, 2, 3].each { |t| refute @counter.throttle?(@key, t), "timestamp #{t}" }
      assert @counter.throttle? @key, 10
      [17, 18, 19].each { |t| refute @counter.throttle?(@key, t), "timestamp #{t}" }
      assert @counter.throttle? @key, 20
    end
  end

  def do_max_requests_minus_one(offset=0)
    [0, 3, 5, 9].map { |t| t + offset }.each do |t|
      refute @counter.throttle?(@key, t), "timestamp #{t}"
    end
  end
end


lib/rack/defense/throttle.rb
module Rack
  class Defense
    class Throttle

      KEY_PREFIX = 'rack-defense'

      attr_accessor :logger
      attr_accessor :name

      def initialize(name, max_requests, time_period, store)
        @name, @max_requests, @time_period = name.to_s, max_requests.to_i, time_period.to_i
        @store = store
      end

      def throttle?(key, timestamp=nil)
        timestamp ||= (Time.now.utc.to_f * 1000).to_i
        @store.eval SCRIPT,
          ["#{KEY_PREFIX}:#{@name}:#{key}"],
          [timestamp, @max_requests, @time_period]
      end

      SCRIPT = <<-LUA_SCRIPT
      local key = KEYS[1]
      local timestamp, max_requests, time_period = tonumber(ARGV[1]), tonumber(ARGV[2]), tonumber(ARGV[3])
      if redis.call('rpush', key, timestamp) < max_requests then
        return false
      else
        return (timestamp - tonumber(redis.call('lpop', key))) <= time_period
      end
      LUA_SCRIPT

      private_constant :SCRIPT

    end
  end
end


lib/rack/defense/throttle_counter.rb
module Rack
  class Defense
    class ThrottleCounter
      KEY_PREFIX = 'rack-defense'

      attr_accessor :name

      def initialize(name, max_requests, time_period, store)
        @name, @max_requests, @time_period = name.to_s, max_requests.to_i, time_period.to_i
        raise ArgumentError, 'name should not be nil or empty' if @name.empty?
        raise ArgumentError, 'max_requests should be greater than zero' unless @max_requests > 0
        raise ArgumentError, 'time_period should be greater than zero' unless @time_period > 0
        @store = store
      end

      def throttle?(key, timestamp=nil)
        timestamp ||= (Time.now.utc.to_f * 1000).to_i
        @store.eval SCRIPT,
          ["#{KEY_PREFIX}:#{@name}:#{key}"],
          [timestamp, @max_requests, @time_period]
      end

      SCRIPT = <<-LUA_SCRIPT
      local key = KEYS[1]
      local timestamp, max_requests, time_period = tonumber(ARGV[1]), tonumber(ARGV[2]), tonumber(ARGV[3])
      local throttle = (redis.call('rpush', key, timestamp) > max_requests) and
                       (timestamp - time_period) <= tonumber(redis.call('lpop', key))
      redis.call('pexpire', key, time_period)
      return throttle
      LUA_SCRIPT

      private_constant :SCRIPT
    end
  end
end


lib/rack/defense.rb
require 'rack'
require 'redis'
require 'delegate'

class Rack::Defense
  autoload :ThrottleCounter, 'rack/defense/throttle_counter'

  class Config
    BANNED_RESPONSE = ->(_) { [403, {'Content-Type' => 'text/plain'}, ["Forbidden\n"]] }
    THROTTLED_RESPONSE = ->(_) { [429, {'Content-Type' => 'text/plain'}, ["Retry later\n"]] }

    attr_accessor :banned_response
    attr_accessor :throttled_response

    attr_reader :bans
    attr_reader :throttles
    attr_reader :ban_callbacks
    attr_reader :throttle_callbacks

    def initialize
      self.banned_response = BANNED_RESPONSE
      self.throttled_response = THROTTLED_RESPONSE
      @throttles, @bans = {}, {}
      @ban_callbacks, @throttle_callbacks = [], []
     end

    def throttle(rule_name, max_requests, period, &block)
      raise ArgumentError, 'rule name should not be nil' unless rule_name
      
      throttles[rule_name] = lambda do |req|  
        max_requests = max_requests.respond_to?(:call) ? max_requests.call(req) : max_requests
        period = period.respond_to?(:call) ? period.call(req) : period

        raise ArgumentError, 'max_requests should be a number' unless max_requests.is_a?(Numeric)
        raise ArgumentError, 'period should be a number' unless period.is_a?(Numeric)

        counter = ThrottleCounter.new(rule_name, max_requests, period, store)
        key = block.call(req)
        key if key && counter.throttle?(key)
      end
    end

    def ban(rule_name, &block)
      raise ArgumentError, 'rule name should not be nil' unless rule_name
      bans[rule_name] = block
    end

    def after_ban(&block)
      ban_callbacks << block
    end

    def after_throttle(&block)
      throttle_callbacks << block
    end

    def store=(value)
      value = Redis.new(url: value) if value.is_a?(String)
      if @store
        @store.__setobj__(value)
      else
        @store = SimpleDelegator.new(value)
      end
    end

    def store
      # Redis.new uses REDIS_URL environment variable by default as URL.
      # See https://github.com/redis/redis-rb
      @store ||= SimpleDelegator.new(Redis.new)
    end
  end

  class << self
    attr_accessor :config

    def setup
      self.config = Config.new
      yield config
    end

    def ban?(req)
      entry = config.bans.find { |_, filter| filter.call(req) }
      matching_rule = entry[0] if entry
      yield config.ban_callbacks, req, matching_rule if matching_rule && block_given?
      matching_rule
    end

    def throttle?(req)
      matching_rules = config.throttles.
          map { |rule_name, filter| [rule_name, filter.call(req)] }.
          select { |e| e[1] }.
          to_h
      yield config.throttle_callbacks, req, matching_rules if matching_rules.any? && block_given?
      matching_rules if matching_rules.any?
    end
  end

  def initialize(app)
    @app = app
  end

  def call(env)
    klass, config = self.class, self.class.config
    req = ::Rack::Request.new(env)

    if klass.ban?(req, &method(:invoke_callbacks))
      config.banned_response.call(env)
    elsif klass.throttle?(req, &method(:invoke_callbacks))
      config.throttled_response.call(env)
    else
      @app.call(env)
    end
  end

  private

  def invoke_callbacks(callbacks, req, rule_data)
    callbacks.each do |callback|
      begin
        callback.call(req, rule_data)
      rescue
        # mute exception
      end
    end
  end
end
